### PR TITLE
perf: Reduce single-system sum and random-key overhead

### PR DESCRIPTION
**Pull request by n-gao**
`segment_sum` now uses a masked `jnp.sum` for one segment, preserving drop/clip behavior and output dtypes. `key_chain` advances its counter on the host and reshapes keys inside the JIT, preserving the random sequence and uint32 wraparound.

Single-segment sums use native JAX accumulation precision: float32 stays float32. Multiple-segment scatters retain their wider accumulator.

N₂ GCMC attempted moves/s on the same node; medians of five 10-second blocks after warmup, float64, eight cycles per dispatch. Initialization and compilation excluded; batch throughput is aggregate.

| Configuration | Before | After | Change |
|---|---:|---:|---:|
| CPU core 1, one system | 2,127 | 2,201 | +3.5% |
| NVIDIA L4, one system | 1,661 | 1,706 | +2.7% |
| NVIDIA L4, 32 systems | 33,139 | 33,926 | +2.4% |

Validation: CPU utility/data/Ewald checks and 125 GPU regression tests passed, plus counter-wraparound checks with x64 disabled on both backends. Ruff, Pyrefly and diff checks passed. Full-energy discrepancies stayed below 3.5e-12 eV; fixed host positions were unchanged.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#334** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/334?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#337** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/337?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#336** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/336?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#335** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/335?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/335?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#333** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/333?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#332** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/332?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#331** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/331?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#330** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/330?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#329** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/329?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#328** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/328?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by felixmusil — APPROVED** [review]

**Comment by felixmusil** [comment]
### Merge activity

* **Sep 18, 9:43 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/335).
* **Sep 18, 11:28 AM UTC**: [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/335) rebased this pull request as part of a merge.
* **Sep 18, 11:49 AM UTC**: @felixmusil merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/335).
